File: src/test/java/dev/hilla/generator/example/CustomTypesEndpoint.java
package dev.hilla.generator.example;

import com.example.library.MyLibraryType;
import dev.hilla.generator.example.entities.MyOtherType;
import dev.hilla.parser.annotations.Endpoint;
import java.util.List;

@Endpoint
public class CustomTypesEndpoint {
  public List<MyCustomType> fromCustomToListOfCustoms(MyCustomType argument) {
    return List.of(argument);
  }

  public MyOtherType fromArrayOfCustomsToCustom(MyOtherType[] argument) {
    return argument[0];
  }

  public MyLibraryType externalLibraryType(MyLibraryType argument) {
    return argument;
  }

  public static class MyCustomType {
    private String field;

    public String getField() {
      return field;
    }

    public void setField(String field) {
      this.field = field;
    }
  }
}

File: src/test/resources/dev/hilla/generator/example/CustomTypesEndpoint.ts
import { MyLibraryType } from '../../../../com/example/library/MyLibraryType.js';
import client from '../../../../connect-client.default.js';
import { MyCustomType } from './CustomTypesEndpoint/MyCustomType.js';
import { MyOtherType } from './entities/MyOtherType.js';
import { EndpointRequestInit } from '@hilla/frontend';

const externalLibraryType = async (argument: MyLibraryType, init?: EndpointRequestInit): Promise<MyLibraryType> => {
    return client.call('dev.hilla.generator.example.CustomTypesEndpoint', 'externalLibraryType', { argument }, init);
};

const fromArrayOfCustomsToCustom = async (argument: MyOtherType[], init?: EndpointRequestInit): Promise<MyOtherType> => {
    return client.call('dev.hilla.generator.example.CustomTypesEndpoint', 'fromArrayOfCustomsToCustom', { argument }, init);
};

const fromCustomToListOfCustoms = async (argument: MyCustomType, init?: EndpointRequestInit): Promise<MyCustomType[]> => {
    return client.call('dev.hilla.generator.example.CustomTypesEndpoint', 'fromCustomToListOfCustoms', { argument }, init);
};

const CustomTypesEndpoint = {
    externalLibraryType,
    fromArrayOfCustomsToCustom,
    fromCustomToListOfCustoms,
};

export default CustomTypesEndpoint;
